File: 133-clone-graph/clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {

        if (node == null) {
            return null;
        }

        // hashmap of form <original node : copy node> to keep track of already copied nodes
        Map<Node, Node> clones = new HashMap<>();

        // add head node to list
        clones.put(node, new Node(node.val));

        // queue to hold next level
        Queue<Node> next = new LinkedList<>();
        next.add(node);

        // bfs traversal
        while (!next.isEmpty()) {
            Node curr = next.remove();

            for (Node neighbor : curr.neighbors) {

                // if neighbor not in map yet, put in map
                if (!clones.containsKey(neighbor)) {
                    clones.put(neighbor, new Node(neighbor.val));

                    // enqueue neighbor
                    next.add(neighbor);
                }

                // add clone of neighbor to neighbor list for clone of curr
                clones.get(curr).neighbors.add(clones.get(neighbor));
            }
        }

        // return head node
        return clones.get(node);
    }
}

File: 133-clone-graph/clone-graph.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution(object):
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """

        if node == None:
            return None

        # map to keep track of cloned nodes
        clones = dict()

        # copy input node
        clones[node] = Node(node.val)

        # queue to hold next nodes to visit
        queue = list()
        queue.append(node)

        # bfs traversal
        while len(queue) > 0:

            # get next node to visit
            curr = queue.pop(0)

            # process neighbors of curr
            for neighbor in curr.neighbors:

                # if neighbor not cloned/visited yet -> clone, add to map, add to queue to visit later
                if neighbor not in clones:
                    clones[neighbor] = Node(neighbor.val)
                    queue.append(neighbor)
                
                # add clone of neighbor to neighbor list of clone of curr
                clones[curr].neighbors.append(clones[neighbor])
        
        return clones[node]
                
